firedrake/assemble.py

```diff
@@ -537,6 +538,10 @@ def get_map(x):
         if needs_cell_facets:
             assert integral_type == "cell"
             extra_args.append(m.cell_to_facets(op2.READ))
+        if pass_layer_arg:
+            import numpy
+            o = base.Arg(op2.Global(1, itspace.layers-2, dtype=numpy.dtype(numpy.int32)))
```

Comment:
I was not 100 percent sure, if this is the right place to generate the argument for the amount of layers?

Reply:
That looks reasonable. You can  just do:

```
c = op2.Global(1, itspace.layers-2, dtype=...)
o = c(op2.READ)
```
I think.

Reply:
This is a simpler fix than I was thinking of (which was to augment kernels saying that they needed this and then wrap it all up in PyOP2).

I think this approach is fine.

Reply:
Your suggestion resulted in a c compiler error, so I just left it as it was.

Reply:
WAH? I don't understand, not least because your version doesn't set an access descriptor for the argument.

Reply:
Which I think should not be allowed, although I see we don't actually check.

Reply:
Ah. I know why, I had not specified a const qualifier in the COFFEE argument declaration.